server/src/main/java/org/elasticsearch/common/lucene/Lucene.java

```diff
@@ -1085,4 +1086,33 @@ public static void scanSeqNosInReader(DirectoryReader directoryReader, long from
             }
         }
     }
+
+    /**
+     * Returns a new Bits which is the union of the given Bits and a DocIdSetIterator.
+     * If the DocIdSetIterator is empty, the same Bits instance will be returned.
+     */
+    public static Bits union(Bits bits, DocIdSetIterator iterator) throws IOException {
+        iterator.nextDoc();
+        if (iterator.docID() == DocIdSetIterator.NO_MORE_DOCS) {
+            return bits;
+        } else {
+            return new Bits() {
+                @Override
+                public boolean get(int index) {
+                    if (bits.get(index)) {
+                        return true;
+                    }
+                    try {
+                        return iterator.advance(index) == index;
```

Comment:
Does this work safely? Javadoc says index must be >= current index, and I am not convinced that this will always be true. But this could certainly be a lack of my knowledge of the actual iterator in use and the usage pattern for the specific query. Thought I would raise this anyway if nothing else then to learn from it.

Reply:
@henningandersen Good catch. We need to check the current docId against the `index` before calling `advance`.